File: tools/bench/ringbuffer_mod_vs_mask.cpp
// 링 버퍼 인덱스 접기: `% cap` 과 `& (cap-1)` 의 실제 차이 (C++).
//
// 컴파일러가 상수 나눗셈을 곱셈+시프트로 바꾸므로 `cap` 이 컴파일 시점 상수면
// 차이가 거의 사라진다. 실제 링 버퍼의 용량은 런타임 값인 경우가 많다.
// 그래서 두 경우를 나눠 잰다 — 상수 cap 과 런타임 cap.
//
// 측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / g++ 13.3.0 -std=c++17 -O2 / 4 코어).
// 빌드·실행: g++ -std=c++17 -O2 tools/bench/ringbuffer_mod_vs_mask.cpp -o /tmp/mm && /tmp/mm

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <vector>
using namespace std;
using namespace std::chrono;

static const int N = 200'000'000;
static const int REPEAT = 7;

// volatile 로 결과를 흘려 보내 최적화로 루프가 통째로 사라지는 것을 막는다.
volatile int sink = 0;

template <class F>
double timeit(F f) {
    auto t0 = steady_clock::now();
    f();
    return duration<double>(steady_clock::now() - t0).count();
}

double med(vector<double> v) {
    sort(v.begin(), v.end());
    return v[v.size() / 2];
}

int main() {
    // 런타임 값. 컴파일러가 상수로 접어 넣지 못하도록 volatile 을 거친다.
    volatile int vcap = 1024;
    int cap = vcap;
    int mask = cap - 1;

    vector<double> tm, tk, tc;
    for (int r = 0; r < REPEAT; r++) {
        tm.push_back(timeit([&] {
            int i = 0;
            for (int k = 0; k < N; k++) i = (i + 1) % cap;   // 런타임 cap
            sink = i;
        }));
        tk.push_back(timeit([&] {
            int i = 0;
            for (int k = 0; k < N; k++) i = (i + 1) & mask;  // 런타임 mask
            sink = i;
        }));
        tc.push_back(timeit([&] {
            int i = 0;
            for (int k = 0; k < N; k++) i = (i + 1) % 1024;  // 컴파일 시점 상수
            sink = i;
        }));
    }

    auto show = [](const char* name, vector<double>& v) {
        double lo = *min_element(v.begin(), v.end());
        double hi = *max_element(v.begin(), v.end());
        printf("%-18s 중앙값 %.3f초  범위 %.3f~%.3f  (%d회, %d회 실행)\n",
               name, med(v), lo, hi, N, REPEAT);
        printf("%-18s 연산당 %.2f ns\n", "", med(v) / N * 1e9);
    };
    show("% cap (런타임)", tm);
    show("& mask (런타임)", tk);
    show("% 1024 (상수)", tc);
    return 0;
}

File: tools/bench/ringbuffer_mod_vs_mask.py
"""링 버퍼 인덱스 접기: `% cap` 과 `& (cap-1)` 의 실제 차이 (Python).

용량이 2의 거듭제곱이면 나머지 연산을 비트 마스크로 바꿀 수 있다.
커널·오디오 링 버퍼가 2의 거듭제곱 용량을 고집하는 근거가 이것인데,
그 근거가 Python 에서도 유효한지는 재 봐야 안다.

측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / CPython 3.13.12 / 4 코어).
실행: python3.13 tools/bench/ringbuffer_mod_vs_mask.py
"""

import statistics
import timeit

CAP = 1024
N = 2_000_000
REPEAT = 7


def bench(stmt, setup):
    ts = timeit.repeat(stmt, setup=setup, number=1, repeat=REPEAT)
    return statistics.median(ts), min(ts), max(ts)


SETUP = f"cap = {CAP}; mask = cap - 1; n = {N}"

MOD = """
i = 0
for _ in range(n):
    i = (i + 1) % cap
"""

MASK = """
i = 0
for _ in range(n):
    i = (i + 1) & mask
"""

# 비교 기준선: 접지 않는 루프. 나머지 연산 자체의 비용이 루프 오버헤드에
# 묻히는지 보려면 루프만 도는 시간을 알아야 한다.
BARE = """
i = 0
for _ in range(n):
    i = i + 1
"""

if __name__ == "__main__":
    for name, stmt in (("bare", BARE), ("%  cap", MOD), ("& mask", MASK)):
        med, lo, hi = bench(stmt, SETUP)
        print(f"{name:8s} 중앙값 {med:.3f}초  범위 {lo:.3f}~{hi:.3f}  ({N:,}회, {REPEAT}회 실행)")
